packages/utils/src/modules-sdk/loaders/container-loader-factory.ts

```diff
@@ -0,0 +1,170 @@
+    repositoryLoader({
+      moduleModels,
+      moduleRepositories,
+      customRepositories: customRepositories ?? {},
+      container,
+    })
+  }
+}
+
+/**
+ * Load the services from the module services object. If a service is not
+ * present a default service will be created for the model.
+ *
+ * @param moduleModels
+ * @param moduleServices
+ * @param container
+ */
+export function loadModuleServices({
+  moduleModels,
+  moduleServices,
+  container,
+}: ServiceLoaderOptions) {
+  const moduleServicesMap = new Map(
+    Object.entries(moduleServices).map(([key, repository]) => [
+      lowerCaseFirst(key),
+      repository,
+    ])
+  )
+
+  // Build default services for all models that are not present in the module services
+  Object.values(moduleModels).forEach((Model) => {
+    const mappedServiceName = lowerCaseFirst(Model.name) + "Service"
+    const finalService = moduleServicesMap.get(mappedServiceName)
+
+    if (!finalService) {
+      moduleServicesMap.set(mappedServiceName, abstractServiceFactory(Model))
+    }
+  })
+
+  const allServices = [...moduleServicesMap]
+
+  allServices.forEach(([key, service]) => {
+    container.register({
+      [lowerCaseFirst(key)]: asClass(service as Constructor<any>).singleton(),
+    })
+  })
+}
+
+/**
+ * Load the repositories from the custom repositories object. If a repository is not
+ * present in the custom repositories object, the default repository will be used from the module repository.
+ * If none are present, a default repository will be created for the model.
+ *
+ * @param moduleModels
+ * @param moduleRepositories
+ * @param customRepositories
+ * @param container
+ */
+export function loadModuleRepositories({
+  moduleModels,
+  moduleRepositories = {},
+  customRepositories,
+  container,
+}: RepositoryLoaderOptions) {
+  const customRepositoriesMap = new Map(
+    Object.entries(customRepositories).map(([key, repository]) => [
+      lowerCaseFirst(key),
+      repository,
+    ])
+  )
+  const moduleRepositoriesMap = new Map(
+    Object.entries(moduleRepositories).map(([key, repository]) => [
+      lowerCaseFirst(key),
+      repository,
+    ])
+  )
+
+  // Build default repositories for all models that are not present in the custom repositories or module repositories
+  Object.values(moduleModels).forEach((Model) => {
+    const mappedRepositoryName = lowerCaseFirst(Model.name) + "Repository"
+    let finalRepository = customRepositoriesMap.get(mappedRepositoryName)
+    finalRepository ??= moduleRepositoriesMap.get(mappedRepositoryName)
+
+    if (!finalRepository) {
+      moduleRepositoriesMap.set(
+        mappedRepositoryName,
+        mikroOrmBaseRepositoryFactory(Model)
+      )
+    }
+  })
+
+  const allRepositories = [...customRepositoriesMap, ...moduleRepositoriesMap]
+
+  allRepositories.forEach(([key, repository]) => {
+    let finalRepository = customRepositoriesMap.get(key)
```

Comment:
i have a bit of a confusion with what a custom repository is and what a overridden repository is (one with an actual file under repositories/*). Would that still be considered here?

Reply:
yes, there is 3 types of repository, the one contains in the @repositories from the modules, the custom one to replace the DAL from the module options and then the on the fly one. If a model does not have a repository in the first 2 then we create one on the fly that we assign to the module repositories map. Does that mane sense